### PR TITLE
NMT combine embedding plus encoder/decoder into one neural module.

### PR DESCRIPTION
**Pull request by ericharper**
This PR create encoder and decoder neural modules for `TransformerEmbedding` and `TransfomerEncoder` / `TransformerDecoder`.

This is making it so that the neural types will be the same across NeMo Transformer, HF BERT, and Megatron BERT encoders.

The idea is to have `EncoderModule` which accepts `input_ids`, ` encoder_mask`, and `num_tokentypes` and outputs `last_hidden_states`.

Then  `DecoderModule` will accept `input_ids`, `decoder_mask`, `encoder_mask` and `encoder_states` and outputs `last_hidden_states`.

This PR also cleans up the relevant configs as the configuration needed to be updated to accept the new model structure.

### CODE REVIEW

File: nemo/collections/nlp/modules/common/decoder_module.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class DecoderModule(NeuralModule):
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        return {
+            "input_ids": NeuralType(('B', 'T'), ChannelType()),
+            "decoder_mask": NeuralType(('B', 'T'), MaskType(), optional=True),
+            "encoder_mask": NeuralType(('B', 'T'), MaskType(), optional=True),
+            "encoder_hidden_states": NeuralType(('B', 'T', 'D'), ChannelType()),
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/decoder_module.py:30]: this should be optional as well. A language model is decoder-only without encoder_hidden_states

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    inner_size: int = MISSING
+
+    # embedding
+    max_sequence_length: int = 512
+    num_token_types: int = 2
+    embedding_dropout: float = 0.0
+    learn_positional_encodings: bool = False
+
+    # transformer
+    num_attention_heads: int = 1
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:39]: this is very important parameter which is related with hidden_size, let's keep it MISSING like hidden_size and not provide a default value for it.

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        self,
+        vocab_size: int,
+        hidden_size: int,
+        num_layers: int,
+        inner_size: int,
+        max_sequence_length: int = 512,
+        num_token_types: int = 2,
+        embedding_dropout: float = 0.0,
+        learn_positional_encodings: bool = False,
+        num_attention_heads: int = 1,
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:63]: let's either not give this default, or set it to 16

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:115]: same here, let's either not give default or use 16

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        mask_future: bool = False,
+        pre_ln: bool = False,
+    ):
+        super().__init__()
+
+        self.vocab_size = vocab_size
+        self.hidden_size = hidden_size
+
+        self.embedding = TransformerEmbedding(
+            vocab_size=vocab_size,
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:77]: please pass here:
```
vocab_size=self.vocab_size
hidden_size=self.hidden_size
```
would make it a little less error-prone when others add changes to these files.

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            hidden_size=hidden_size,
+            max_sequence_length=max_sequence_length,
+            num_token_types=num_token_types,
+            embedding_dropout=embedding_dropout,
+            learn_positional_encodings=learn_positional_encodings,
+        )
+
+        self.encoder = TransformerEncoder(
+            hidden_size=hidden_size,
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:86]: please pass here:
```
vocab_size=self.vocab_size
hidden_size=self.hidden_size
```
would make it a little less error-prone when others add changes to these files.

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        hidden_act: str = 'relu',
+        pre_ln: bool = False,
+    ):
+        super().__init__()
+
+        self.hidden_size = hidden_size
+        self.max_sequence_length = max_sequence_length
+
+        self.embedding = TransformerEmbedding(
+            vocab_size=vocab_size,
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:128]: please pass here:
```
vocab_size=self.vocab_size
hidden_size=self.hidden_size
```
would make it a little less error-prone when others add changes to these files.

File: nemo/collections/nlp/modules/common/transformer/transformer.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            hidden_size=hidden_size,
+            max_sequence_length=max_sequence_length,
+            num_token_types=num_token_types,
+            embedding_dropout=embedding_dropout,
+            learn_positional_encodings=learn_positional_encodings,
+        )
+
+        self.decoder = TransformerDecoder(
+            hidden_size=hidden_size,
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer.py:137]: please pass here:
```
vocab_size=self.vocab_size
hidden_size=self.hidden_size
```
would make it a little less error-prone when others add changes to these files.

**Review by okuchaiev — CHANGES_REQUESTED** [review]
looks good to me, just few small comments.

File: nemo/collections/nlp/modules/common/decoder_module.py (hunk 1 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/decoder_module.py:30]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 2 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:39]: I think it should be set to `MISSING`. Will update the configs.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 3 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:63]: Removed default.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 4 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:115]: Removed default.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 5 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:77]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 6 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:86]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 7 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:128]: Done.

File: nemo/collections/nlp/modules/common/transformer/transformer.py (hunk 8 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/transformer.py:137]: Done.

**Pull request merged**